File: 1582-design-browser-history/design-browser-history.cpp
class BrowserHistory {
public:
    stack<string> backStack;
    stack<string> forwardStack;
    string current;
    BrowserHistory(string homepage) {
        current = homepage;
    }
    
    void visit(string url) {
        while(!forwardStack.empty())
        {
            forwardStack.pop();
        }
        backStack.push(current);
        current = url;
    }
    
    string back(int steps) {
        while(steps > 0 && !backStack.empty())
        {
            forwardStack.push(current);
            current = backStack.top();
            backStack.pop();
            steps--;
        }
        return current;
    }
    
    string forward(int steps) {
        while(steps > 0 && !forwardStack.empty())
        {
            backStack.push(current);
            current = forwardStack.top();
            forwardStack.pop();
            steps--;
        }
        return current;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: 1582-design-browser-history/design-browser-history.cs
public class BrowserHistory
{
    public class Node
    {
        public Node prev;
        public string val;
        public Node next;

        public Node(Node prev = null, string val = "", Node next = null)
        {
            this.prev = prev;
            this.val = val;
            this.next = next;
        }
    }

    private Node currentPage;

    public BrowserHistory(string homepage)
    {
        currentPage = new Node(null, homepage, null);
    }

    public void Visit(string url)
    {
        currentPage.next = null;

        Node newNode = new Node(currentPage, url, null);
        currentPage.next = newNode;
        currentPage = newNode;
    }

    public string Back(int steps)
    {
        while (steps > 0 && currentPage.prev != null)
        {
            currentPage = currentPage.prev;
            steps--;
        }
        return currentPage.val;
    }

    public string Forward(int steps)
    {
        while (steps > 0 && currentPage.next != null)
        {
            currentPage = currentPage.next;
            steps--;
        }
        return currentPage.val;
    }
}